### PR TITLE
[SPARK-41349][CONNECT][PYTHON] Implement DataFrame.hint

### PR DESCRIPTION
**Pull request by dengziming**
### What changes were proposed in this pull request?
Implement DataFrame.hint for pyspark

### Why are the changes needed?
For API coverage

### Does this PR introduce _any_ user-facing change?
new APIs

### How was this patch tested?
existing python tests.

### CODE REVIEW

File: python/pyspark/sql/connect/proto_converter.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    elif isinstance(value, datetime.datetime):
+        literal.timestamp = TimestampType().toInternal(value)
+    elif isinstance(value, datetime.date):
+        literal.date = DateType().toInternal(value)
+    elif isinstance(value, datetime.timedelta):
+        interval = DayTimeIntervalType().toInternal(value)
+        assert interval is not None
+        literal.day_time_interval = int(interval)
+    else:
+        return None
```
Comment (dengziming) [inline python/pyspark/sql/connect/proto_converter.py:60]: Not sure is this the best practice, let's improve it if there is a better way.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 2]
```diff
@@ -305,7 +305,11 @@ class SparkConnectPlanner(session: SparkSession) {
   }
 
   private def transformHint(rel: proto.Hint): LogicalPlan = {
-    val params = rel.getParametersList.asScala.map(toCatalystValue).toSeq
+    val params = rel.getParametersList.asScala.map(toCatalystValue).toSeq.map{
+      case name: String => UnresolvedAttribute(name)
+      case l: Long => l.toInt
```
Comment (amaliujia) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:310]: I am a bot confused here: why need this convert?

Comment (dengziming) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:310]: Python only has int type and has no long type, so we can't pass in a int literal since we always assumes the largest type.

Comment (HyukjinKwon) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:311]: sorry if I am being dumb but why do we need pass a int literal?

Comment (dengziming) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:311]: Hi @HyukjinKwon , using a long value will fail scala pattern match here since it will only accept a int value:

https://github.com/apache/spark/blob/c78f935185cbfae2e45adb393def0bd389a89c7e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala#L242-L249

**Comment by amaliujia** [comment]
Looks like the CI is failing on 
1. python/pyspark/sql/tests/connect/test_connect_basic.py.test_fill_na doe not pass
2. cannot pass ./dev/lint-scala

**Comment by zhengruifeng** [comment]
if `test_fill_na` is irrelevant to this PR, I guess rebasing may help

File: python/pyspark/sql/connect/column.py [hunk 3]
```diff
@@ -18,26 +18,18 @@
...
     from pyspark.sql.connect.client import SparkConnectClient
     import pyspark.sql.connect.proto as proto
 
 
-JVM_INT_MIN = -(1 << 31)
-JVM_INT_MAX = (1 << 31) - 1
-JVM_LONG_MIN = -(1 << 63)
-JVM_LONG_MAX = (1 << 63) - 1
-
-
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/column.py:40]: Why is all of this removed?

File: python/pyspark/sql/connect/types.py [hunk 4]
```diff
@@ -41,6 +41,15 @@
     BooleanType,
     NullType,
 )
+import decimal
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/types.py:44]: While I generally agree with the refactoring, I think it would be good to split this into two PRs. One that does the refactoring and then this one. Otherwise this just combines unrelated changes into one PR.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 2 above)
Comment (zhengruifeng) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala]: after https://github.com/apache/spark/pull/39047, I think you can now create a literal int from the client side

Comment (dengziming) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala]: Yes, Thank for pointing out this.

**Comment by dengziming** [comment]
ping @zhengruifeng @amaliujia @grundprinzip

File: python/pyspark/sql/connect/dataframe.py [hunk 5]
```diff
@@ -875,6 +875,30 @@ def to_jcols(
...
+        that one of the plan can be broadcasted: `df1.join(df2.hint("broadcast"))`
+
+        .. versionadded:: 3.4.0
+
+        Parameters
+        ----------
+        name: str
+            the name of the hint, for example, "broadcast", "SHUFFLE_MERGE" and "shuffle_hash".
+        params: tuple
+            the parameters of the hint
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/dataframe.py:890]: I know that the documentation is most likeley directly from PySpark, but I'm wondering if we can add more context around what types can the params have? If I read through the code it can be `any` here but later only `Optional[Union[str, int]]`?

File: python/pyspark/sql/connect/plan.py [hunk 6]
```diff
@@ -343,6 +343,51 @@ def _repr_html_(self) -> str:
...
+
+    def _convert_value(self, v: Any) -> proto.Expression.Literal:
+        value = proto.Expression.Literal()
+        if v is None:
+            value.null = True
+        elif isinstance(v, int):
+            value.integer = v
+        else:
+            value.string = v
+        return value
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/plan.py:362]: This code has weird error behavior if v is not `None` or `int`. If for example, I were to assign a float, I would receive an error message from protobuf that is not actionable for the user. I think it would be good to either use the existing Python to Literal conversion code that we have or throw an exception.

File: python/pyspark/sql/tests/connect/test_connect_basic.py [hunk 7]
```diff
@@ -829,6 +829,13 @@ def test_with_columns(self):
             .toPandas(),
         )
 
+    def test_hint(self):
+        # SPARK-41349: Test hint
```
Comment (grundprinzip) [inline python/pyspark/sql/tests/connect/test_connect_basic.py:833]: Please add additional tests for:

  * unsupported param types
  * unsupported hint name
  * invalid combination of hint and param

Please check if there is additional coverage needed.

**Review by grundprinzip — COMMENTED** [review]
Functional code looks good, please address the comments with regard to error handling and tests. 

Thanks for your contribution!

File: python/pyspark/sql/connect/dataframe.py (hunk 5 above)
Comment (dengziming) [reply python/pyspark/sql/connect/dataframe.py]: Yeah, It's worth improving the docs of the parameters.

File: python/pyspark/sql/connect/plan.py (hunk 6 above)
Comment (dengziming) [reply python/pyspark/sql/connect/plan.py]: I improved the error handing logic here, and there are 4 occurrence of this similar logic, I'm planing to refactor it to reuse the existing code.

File: python/pyspark/sql/tests/connect/test_connect_basic.py (hunk 7 above)
Comment (dengziming) [reply python/pyspark/sql/tests/connect/test_connect_basic.py:836]: Good catch, I added 4 more test cases for it.

**Comment by dengziming** [comment]
@grundprinzip Thank you for your reviews, I have resolved these comments.

**Review by dengziming — COMMENTED** [review]
@grundprinzip  Thank you for your reviews, I have resolved these comments.

File: python/pyspark/sql/connect/plan.py [hunk 8]
```diff
@@ -343,6 +343,53 @@ def _repr_html_(self) -> str:
...
+
+    def plan(self, session: "SparkConnectClient") -> proto.Relation:
+        assert self._child is not None
+        plan = proto.Relation()
+        plan.hint.input.CopyFrom(self._child.plan(session))
+        plan.hint.name = self.name
+        for v in self.params:
+            param = proto.Expression.Literal()
+            param.CopyFrom(self._convert_value(v))
+            plan.hint.parameters.append(param)
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/plan.py:374]:
```suggestion
            plan.hint.parameters.append(self._convert_value(v))
```

**Review by grundprinzip — APPROVED** [review]
One minor improvement, otherwise LGTM.

**Comment by grundprinzip** [comment]
@HyukjinKwon @zhengruifeng please take a look.

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
merged into master, thank you @dengziming for working on it!

**Comment by amaliujia** [comment]
many thanks @dengziming to tackle this work!
